### PR TITLE
Variable Sets redux --> Formatters

### PR DESCRIPTION
**Pull request by kitbs**
I've had a go at reimplementing Formatters based on your solution.

I experimented with `get_class_methods()` and it does return `protected` if not `private` methods. I went with the reflection approach to only include public and zero-argument methods, to be totally sure.

I added validation back in, and some helper functions to work with the variables before they are passed to `render()`. You can also now pass a `Formatter` instance to `render()` as well as an array.

I haven't put anything in the README yet, in case things change as we discuss it, but basically this is the API:

### Basic formatter example

```php
use Stub\Formatter;
use Illuminate\Support\Str;

class MyFormatter extends Formatter
{
    public function plural()
    {
        return Str::plural($this->name);
    }
}
```

### Construct a formatter and retrieve variables with various methods

```php
$formatter = new MyFormatter(['name' => 'User']);

/*
MyFormatter Object
(
    [variables] => Array
        (
            [name] => User
        )
)
*/

```
#### Get all the variables and computed variables
```php
$variables = $formatter->all(); // Call `all()` method
$variables = $formatter(); // Invoke the class

/*
Array
(
    [name] => User
    [plural] => Users
)
*/
```
#### Get a single non-computed variable after the formatter is constructed
```php
$formatter->get('name'); // `get` method
$formatter->name; // Magic `__get`
$formatter->name(); // Magic `__call`, if `name()` method is not already defined
```
Note: `get('name')` and `name` will always return the original variable. If no public method `name()` is defined, this will also magically return the original variable, but if you add a `name()` method to modify the value of the variable, the modified variable will be returned.

#### Get a single computed variable after the formatter is constructed
```php
$formatter->plural();
```
Note: Computed variables can only be retrieved by their own methods, not `get()` or magic methods.

#### Add a single variable after the formatter is constructed
```php
$formatter->set('package', 'vendor/package');
```

#### Remove a single variable after the formatter is constructed
```php
$formatter->unset('package');
```

#### Merge an array of variables after the formatter is constructed
```php
$formatter->merge([
    'package' => 'vendor/package',
    'additional' => 'extra',
);
```

#### Replace all variables after the formatter is constructed
```php
$formatter->replace([
    'package' => 'vendor/package',
    'additional' => 'extra',
);
```

### Make a formatter and return original and computed variables at once

```php
$variables = MyFormatter::make(['name' => 'User']);

/*
Array
(
    [name] => User
    [plural] => Users
)
*/
```

### CODE REVIEW

**Comment by kitbs** [comment]
I also haven't updated the console commands with an option to pass in a formatter, but I think this would be a good idea.

**Comment by dillingham** [comment]
@kitbs This looks great! I'll review it further in a bit.

File: tests/FormatterTest.php [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+        ], $variables);
+    }
+
+    public function testFormatterVariableMerge()
+    {
+        $formatter = new Formatters\Example1(['name' => 'User']);
+
+        $this->assertFalse($formatter->has('package'), 'Failed asserting that variable \'package\' does not exist.');
+
+        $formatter->merge([
```
Comment (dillingham) [inline tests/FormatterTest.php]: Can you help me understand this line,  trying to imagine an instance where the coder would pass an array to format but not have all the items. Wondering if that check would be pre-stub or is this more of a test to verify internal behavior?

File: src/Formatter.php [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+        return $this;
+    }
+
+    /**
+     * Replace existing variables with new variables.
+     *
+     * @param string[] $variables
+     * @return $this
+     */
+    final public function replace(array $variables)
```
Comment (dillingham) [inline src/Formatter.php]: While this does make working with `Formatter`s on their own pretty cool.. wondering if its icing on the cake that we might not need for the current goal of `computed variables`.  Is it something you think we need?

File: src/Formatter.php [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+        return $this;
+    }
+
+    /**
+     * Remove a variable.
+     *
+     * @param string $variable
+     * @return $this
+     */
+    final public function unset(string $variable)
```
Comment (dillingham) [inline src/Formatter.php]: Sort of along the lines with `replace()`, I think `set()` `get()` `has()`  `all()`  have strong reasons to exist. They making the code cleaner and/or perform the needed tasks for the feature. I think these extra methods make `Formatter` as a whole something I'd want to pull into a project haha but maybe we could thin it out for our current objective. What are your thoughts?

**Comment by dillingham** [comment]
Those are my only thoughts for this. Overall I think this code is awesome and very well tested /  professional and am looking forward to adding this feature to the readme haha

**Comment by kitbs** [comment]
Yeah, I totally agree that maybe some of the methods are not essential for the basic use case we're trying to meet.

If I'm clutching at straws, a flimsy use case for this might be one of the Q&A console jobs, where you build up the variables gradually. But granted, that could all be done externally to the class just with a couple of simple `$array[] = $value;`s before constructing the `Formatter`.

I probably gota little carried away, but was just feeling a little anxious about making a "closed box" for the variables without any way to modify them after constructing. But perhaps it should really be an immutable object?

If you think it's too much icing and not enough cake, I think the approach would be:
- Move the logic of `merge()` back into the constructor
- Remove `__set()`, `set()`,  `unset()`, `replace()` and `merge()`
- Keep `__get()`, `get()` and `has()`

 It certainly cuts down on the complexity of the README, I'll give you that! 😆

**Pull request merged**